The aggregation count can't be applied to a CTE

**Describe the bug**
The count(*) syntax can't be applied to a CTE.
It's an issue responded to by the Discord community https://discord.com/channels/1227143286951514152/1285568489414856815/1285568647775129682
**To Reproduce**

Given the request body for preview API
```
{
  "manifest": {
    "catalog": "wrenai",
    "schema": "public",
    "models": [
      {
        "name": "customers",
        "refSql": "select * from main.customers",
        "columns": [
          {
            "name": "City",
            "type": "VARCHAR",
            "isCalculated": false,
            "notNull": false
          },
          {
            "name": "Id",
            "type": "VARCHAR",
            "isCalculated": false,
            "notNull": false
          },
          {
            "name": "State",
            "type": "VARCHAR",
            "isCalculated": false,
            "notNull": false
          }
        ],
        "primaryKey": "Id"
      }
    ]
  },
  "sql": "with t1 as (select * from customers) select count(*) from t1"
}
```
Then, got the error message
```
Caused by: java.lang.IllegalArgumentException: source node name not found: Table{t1, Optional.empty}
	at io.wren.base.sqlrewrite.WrenSqlRewrite.lambda$apply$5(WrenSqlRewrite.java:102)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at io.wren.base.sqlrewrite.WrenSqlRewrite.lambda$apply$8(WrenSqlRewrite.java:102)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at io.wren.base.sqlrewrite.WrenSqlRewrite.apply(WrenSqlRewrite.java:100)
	at io.wren.base.sqlrewrite.WrenSqlRewrite.apply(WrenSqlRewrite.java:83)
	at io.wren.base.sqlrewrite.WrenPlanner.rewrite(WrenPlanner.java:51)
	at io.wren.base.sqlrewrite.WrenPlanner.rewrite(WrenPlanner.java:43)
	at io.wren.main.PreviewService.lambda$dryRun$2(PreviewService.java:129)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
```

**Expected behavior**
The SQL should work.

--- a/wren-base/src/main/java/io/wren/base/sqlrewrite/analyzer/ExpressionAnalyzer.java
+++ b/wren-base/src/main/java/io/wren/base/sqlrewrite/analyzer/ExpressionAnalyzer.java
@@ -36,6 +36,7 @@ import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 
 import static java.util.Objects.requireNonNull;
 
@@ -140,7 +141,7 @@ public class ExpressionAnalyzer
         @Override
         protected Void visitSubqueryExpression(SubqueryExpression node, Void context)
         {
-            StatementAnalyzer.analyze(analysis, node.getQuery(), sessionContext, wrenMDL);
+            StatementAnalyzer.analyze(analysis, node.getQuery(), sessionContext, wrenMDL, Optional.of(scope));
             return null;
         }
 

--- a/wren-base/src/main/java/io/wren/base/sqlrewrite/analyzer/StatementAnalyzer.java
+++ b/wren-base/src/main/java/io/wren/base/sqlrewrite/analyzer/StatementAnalyzer.java
@@ -394,7 +394,6 @@ public final class StatementAnalyzer
         private void analyzeSelectSingleColumn(SingleColumn singleColumn, Scope scope, ImmutableList.Builder<Expression> outputExpressions)
         {
             outputExpressions.add(singleColumn.getAlias().map(name -> (Expression) name).orElse(singleColumn.getExpression()));
-            // TODO: handle when singleColumn is a subquery
             ExpressionAnalysis expressionAnalysis = analyzeExpression(singleColumn.getExpression(), scope);
 
             if (expressionAnalysis.isRequireRelation()) {

--- a/wren-base/src/main/java/io/wren/base/sqlrewrite/analyzer/StatementAnalyzer.java
+++ b/wren-base/src/main/java/io/wren/base/sqlrewrite/analyzer/StatementAnalyzer.java
@@ -83,7 +83,12 @@ public final class StatementAnalyzer
 
     public static Scope analyze(Analysis analysis, Statement statement, SessionContext sessionContext, WrenMDL wrenMDL)
     {
-        Scope queryScope = new Visitor(sessionContext, analysis, wrenMDL).process(statement, Optional.empty());
+        return analyze(analysis, statement, sessionContext, wrenMDL, Optional.empty());
+    }
+
+    public static Scope analyze(Analysis analysis, Statement statement, SessionContext sessionContext, WrenMDL wrenMDL, Optional<Scope> scope)
+    {
+        Scope queryScope = new Visitor(sessionContext, analysis, wrenMDL).process(statement, scope);
         // add models directly used in sql query
         analysis.addModels(
                 wrenMDL.listModels().stream()
@@ -244,7 +249,7 @@ public final class StatementAnalyzer
                     else {
                         SingleColumn singleColumn = (SingleColumn) selectItem;
                         String name = singleColumn.getAlias().map(Identifier::getValue)
-                                .or(() -> Optional.ofNullable(QueryUtil.getQualifiedName(singleColumn.getExpression()).getSuffix()))
+                                .or(() -> Optional.ofNullable(QueryUtil.getQualifiedName(singleColumn.getExpression())).map(QualifiedName::getSuffix))
                                 .orElse(singleColumn.getExpression().toString());
                         if (scope.isPresent()) {
                             Optional<Field> fieldOptional = scope.get().getRelationType().resolveAnyField(QueryUtil.getQualifiedName(singleColumn.getExpression()));
@@ -397,16 +402,18 @@ public final class StatementAnalyzer
                         .orElseThrow(() -> new IllegalArgumentException("count(*) should have a followed source"));
 
                 // collect only the source node that is a table for generating the required column for models
-                DefaultTraversalVisitor<Void> visitor = new DefaultTraversalVisitor<>()
+                DefaultTraversalVisitor<Scope> visitor = new DefaultTraversalVisitor<>()
                 {
                     @Override
-                    protected Void visitTable(Table node, Void scope)
+                    protected Void visitTable(Table node, Scope scope)
                     {
-                        analysis.addRequiredSourceNode(node);
+                        if (scope.getNamedQuery(node.getName().getSuffix()).isEmpty()) {
+                            analysis.addRequiredSourceNode(node);
+                        }
                         return null;
                     }
                 };
-                visitor.process(source, null);
+                visitor.process(source, scope);
             }
         }
 
@@ -570,7 +577,7 @@ public final class StatementAnalyzer
         @Override
         protected Scope visitTableSubquery(TableSubquery node, Optional<Scope> scope)
         {
-            return Optional.ofNullable(analyze(analysis, node.getQuery(), sessionContext, wrenMDL))
+            return Optional.ofNullable(analyze(analysis, node.getQuery(), sessionContext, wrenMDL, scope))
                     .map(value -> createAndAssignScope(node, scope, value))
                     .orElseGet(() -> Scope.builder().parent(scope).build());
         }

--- a/wren-tests/src/test/java/io/wren/testing/duckdb/TestWrenWithDuckDB.java
+++ b/wren-tests/src/test/java/io/wren/testing/duckdb/TestWrenWithDuckDB.java
@@ -232,6 +232,12 @@ public class TestWrenWithDuckDB
 
         queryResultDto = query(manifest, "with t1 as (select * from Orders) select * from (select count(*) from t1) s1");
         assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(15000);
+
+        queryResultDto = query(manifest, "with t1 as (select * from Orders) select 15000 = (select count(*) from t1)");
+        assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(true);
+
+        queryResultDto = query(manifest, "with t1 as (select * from Orders) select 1 from Orders where 15000 = (select count(*) from t1)");
+        assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(1);
     }
 
     @Test

--- a/wren-tests/src/test/java/io/wren/testing/duckdb/TestWrenWithDuckDB.java
+++ b/wren-tests/src/test/java/io/wren/testing/duckdb/TestWrenWithDuckDB.java
@@ -204,7 +204,7 @@ public class TestWrenWithDuckDB
     }
 
     @Test
-    public void testCountJoin()
+    public void testCount()
     {
         QueryResultDto queryResultDto = query(manifest, "select count(*) from Orders a");
         assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(15000);
@@ -216,6 +216,21 @@ public class TestWrenWithDuckDB
         assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(22500000);
 
         queryResultDto = query(manifest, "select count(*) from Orders a JOIN Customer b ON a.custkey = b.custkey");
+        assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(15000);
+
+        queryResultDto = query(manifest, "select count(*) from (select * from Orders) t1");
+        assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(15000);
+
+        queryResultDto = query(manifest, "select count(*) from (Orders a JOIN Customer b ON a.custkey = b.custkey) t1");
+        assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(15000);
+
+        queryResultDto = query(manifest, "with t1 as (select * from Orders) select count(*) from t1");
+        assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(15000);
+
+        queryResultDto = query(manifest, "with t1 as (select * from Orders), t2 as (select count(*) from t1) select * from t2");
+        assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(15000);
+
+        queryResultDto = query(manifest, "with t1 as (select * from Orders) select * from (select count(*) from t1) s1");
         assertThat(queryResultDto.getData().getFirst()[0]).isEqualTo(15000);
     }
 

